flashcard.1
.TH FLASHCARD 1 FLASHCARD\-1.0
.SH NAME
flashcard \- cli flashcard-based learning system based on the leitner system
.SH SYNOPSIS
.B flashcard
.P
.B flashcard
.B list
.IR deck-name
.P
.B flashcard
.B get
.RB [ int | recto | verso ]
.IR card-name
.P
.B flashcard
.B edit
.BR recto | verso
.IR card-name
.IR value
.RB | 
.B (from
.IR file)
.P
.B flashcard
.B edit
.IR card-name
.IR recto-value
.RB |
.B (from
.IR file)
.IR verso-value
.RB |
.B (from
.IR file)
.P
.B flashcard
.IR deck-name 
.RB [ force ]
.SH DESCRIPTION
.B flashcard
is a cli flashcard-based learning system based on the Leitner system and inspired by the Anki application.
.P
When you study a card,
it will choose when to show you the card again based on your performance.
.P
It calculate when to show you the card by * 2, * 4 or / 2
the amount of day you waited to get this card and add it to the current day.
.P
You can modifie of create new deck/card manualy in /usr/local/flashcard by default unless config changed by user.
.SH USAGE
.TP
.B flashcard
.br
With no arguments passed,
flashcard will by default show a tree of your decks and categories
.TP
.B flashcard list dico
List all the card in the dico deck
.TP
.B flashcard get recto dico/grass
.br
Return the recto of grass card in dico deck.
.br
Will do the same with verso and int,
int means interval,
it contain the number of day to wait and the date the card is unlocked
.TP
.B flashcard  edit verso dico/grass "'litle green thing that we need to touch more'"
.TQ
.B flashcard  edit recto dico/grass from ./grass-definition.txt
Change or create the recto or verso face of grass card of dico deck with specified value or from a file
.TP
.B flashcard edit dico/grass "'grass noun /ɡrɑːs/'" "grass definition"
.TQ
.B flashcard edit dico/grass "'grass noun /ɡrɑːs/'" from ./grass-definition.txt
.TQ
.B flashcard edit dico/grass from ./grass-info.txt "grass definition"
.TQ
.B flashcard edit dico/grass from ./grass-info.txt from ./grass-definition.txt
Change or create first the recto and then the verso with specified value or from file.
Source can be different for recto and verso 
.TP
.B flashcard dico [force]
Will play the cards in dico that are unlocked for today and will update there new date.
.br
force is optional, it will force playing all the card even if locked.


flashcard.sh
#!/bin/bash
# TODO : nice feature to have
# reverse / re-reverse mode :
#   show verso first and try to remeber recto
#   switch form reverse mode to normale mode randomly

carddir=/usr/local/flashcard

first_interval="1"
step="2"

bold="\033[1m"
underline="\033[4m"
nocolor="\033[0m"
clear="\033[2J\033[0;0H"
altscreen="\033[?1049h\033[H"
exit_altscreen="\033[?1049l"

[ -e $carddir ] || {
  echo "flashcard dir do not exist, creating it"
  sudo mkdir -p $carddir
  sudo chown $USER:$USER $carddir
}

case $1 in
  "")
    tree -d $carddir
    ;;

  "list")
    ls $carddir/$2 | sed -r "s/\.(int|recto|verso)$//g" | uniq
    ;;

  "get")
    case $2 in
      "recto" | "verso" | "int")
        cat $carddir/$3.$2
        ;;
      *)
        echo -e $bold$underline"interval:"$nocolor
        cat $carddir/$2.int

        echo -e $bold$underline"recto:"$nocolor
        cat $carddir/$2.recto

        echo -e $bold$underline"verso:"$nocolor
        cat $carddir/$2.verso
        ;;
    esac
    ;;

  "edit")
    case $2 in
      "recto")
        mkdir -p $carddir/${3%/*}
        if [[ $4 == "from" ]]; then
          cp $5 $carddir/$3.recto
        else
          echo $4 > $carddir/$3.recto
        fi

        [ -f $carddir/$3.int ] || {
          echo $first_interval > $carddir/$3.int
          date "+%F" >> $carddir/$3.int
        }
        ;;

      "verso")
        mkdir -p $carddir/${3%/*}
        if [[ $4 == "from" ]]; then
          cp $5 $carddir/$3.verso
        else
          echo $4 > $carddir/$3.verso
        fi

        [ -f $carddir/$3.int ] || {
          echo $first_interval > $carddir/$3.int
          date "+%F" >> $carddir/$3.int
        }
        ;;

      *)
        mkdir -p $carddir/${2%/*}
        if [[ "$3" == "from" && "$5" == "from" ]]; then
          cp $4 $carddir/$2.recto
          cp $6 $carddir/$2.verso
        elif [[ "$3" == "from" ]]; then
          cp $4 $carddir/$2.recto
          echo $5 > $carddir/$2.verso
        elif [[ "$4" == "from" ]]; then
          echo $3 > $carddir/$2.recto
          cp $5 $carddir/$2.verso
        else
          echo $3 > $carddir/$2.recto
          echo $4 > $carddir/$2.verso
        fi

        [ -f $carddir/$2.int ] || {
          echo $first_interval > $carddir/$2.int
          date "+%F" >> $carddir/$2.int
        }
        ;;
      esac
    ;;

  *)
    cd $carddir/$1
    cards=$(ls | sed -r "s/\.(int|recto|verso)$//g" | uniq | shuf)

    echo -e $altscreen

    nocard="true"
    for card in $cards; do
      date=$(sed -n '2 p' $card.int)
      [[ $2 != "force" && $(date "+%F") < $date ]] && continue

      nocard="false"
      interval=$(sed -n '1 p' $card.int)
      [ $interval -eq 0 ] && interval=1

      echo "card $card:"
      cat $card.recto
      echo -e "\npress enter to get the verso or enter quit to quit."

      read quit
      [[ $quit == "quit" ]] && exit

      echo "verso:"
      cat $card.verso

      while [[ $valid != "true" ]]; do
        echo -e "\n(r)Reset - (h)Hard - (e)Easy - (s)Super Easy - (q)Quit"
        read reponse

        case $reponse in
          "q" | "quit")
            exit
            ;;
          "r" | "reset")
            echo $first_interval > $card.int
            date "+%F" >> $card.int
            valid="true"
            ;;
          "h" | "hard")
            interval=$(expr $interval \/ $step)
            echo $interval > $card.int
            date "+%F" -d "+$interval day" >> $card.int
            valid="true"
            ;;
          "e" | "easy")
            interval=$(expr $interval \* $step)
            echo $interval > $card.int
            date "+%F" -d "+$interval day" >> $card.int
            valid="true"
            ;;
          "s" | "super easy")
            interval=$(expr $interval \* $step \* 2)
            echo $interval > $card.int
            date "+%F" -d "+$interval day" >> $card.int
            valid="true"
            ;;
          *)
            echo "value not valid"
            ;;
        esac
      done
      valid="false"

      echo -e $clear
    done

    echo -e $exit_altscreen
    ;;
esac

[[ "$nocard" == "true" ]] && echo "no card to learn for today"
